agent/bls.mjs

```diff
@@ -0,0 +1,3130 @@
+        typeof joNode.imaInfo.BLSPublicKey1 === "string" &&
+        joNode.imaInfo.BLSPublicKey1.length > 0 &&
+        "BLSPublicKey2" in joNode.imaInfo &&
+        typeof joNode.imaInfo.BLSPublicKey2 === "string" &&
+        joNode.imaInfo.BLSPublicKey2.length > 0 &&
+        "BLSPublicKey3" in joNode.imaInfo &&
+        typeof joNode.imaInfo.BLSPublicKey3 === "string" &&
+        joNode.imaInfo.BLSPublicKey3.length > 0
+    ) {
+        return {
+            BLSPublicKey0: joNode.imaInfo.BLSPublicKey0,
+            BLSPublicKey1: joNode.imaInfo.BLSPublicKey1,
+            BLSPublicKey2: joNode.imaInfo.BLSPublicKey2,
+            BLSPublicKey3: joNode.imaInfo.BLSPublicKey3
+        };
+    }
+    return null;
+}
+
+function discoverCommonPublicKey( joSChainNetworkInfo ) {
+    const imaState = state.get();
+    joSChainNetworkInfo = joSChainNetworkInfo || imaState.joSChainNetworkInfo;
+    const jarrNodes = joSChainNetworkInfo.network;
+    for( let i = 0; i < jarrNodes.length; ++i ) {
+        const joNode = jarrNodes[i];
+        if( joNode && "imaInfo" in joNode && typeof joNode.imaInfo === "object" &&
+            "commonBLSPublicKey0" in joNode.imaInfo &&
+            typeof joNode.imaInfo.commonBLSPublicKey0 === "string" &&
+            joNode.imaInfo.commonBLSPublicKey0.length > 0 &&
+            "commonBLSPublicKey1" in joNode.imaInfo &&
+            typeof joNode.imaInfo.commonBLSPublicKey1 === "string" &&
+            joNode.imaInfo.commonBLSPublicKey1.length > 0 &&
+            "commonBLSPublicKey2" in joNode.imaInfo &&
+            typeof joNode.imaInfo.commonBLSPublicKey2 === "string" &&
+            joNode.imaInfo.commonBLSPublicKey2.length > 0 &&
+            "commonBLSPublicKey3" in joNode.imaInfo &&
+            typeof joNode.imaInfo.commonBLSPublicKey3 === "string" &&
+            joNode.imaInfo.commonBLSPublicKey3.length > 0
+        ) {
+            return {
+                commonBLSPublicKey0: joNode.imaInfo.commonBLSPublicKey0,
+                commonBLSPublicKey1: joNode.imaInfo.commonBLSPublicKey1,
+                commonBLSPublicKey2: joNode.imaInfo.commonBLSPublicKey2,
+                commonBLSPublicKey3: joNode.imaInfo.commonBLSPublicKey3
+            };
+        }
+    }
+    return null;
+}
+
+function hexPrepare( strHex, isInvertBefore, isInvertAfter ) {
+    if( isInvertBefore == undefined )
+        isInvertBefore = true;
+    if( isInvertAfter == undefined )
+        isInvertAfter = true;
+    let arrBytes = imaUtils.hexToBytes( strHex );
+    if( isInvertBefore )
+        arrBytes = imaUtils.invertArrayItemsLR( arrBytes );
+    arrBytes = imaUtils.bytesAlignLeftWithZeroes( arrBytes, 32 );
+    if( isInvertAfter )
+        arrBytes = imaUtils.invertArrayItemsLR( arrBytes );
+    return arrBytes;
+}
+
+function s2ha( s ) {
+    const strU256 = owaspUtils.ethersMod.ethers.utils.id( s );
+    return hexPrepare( strU256, true, true );
+}
+
+function a2ha( arrBytes ) {
```

Comment:
Please rename to arrayToKeccak256 or similar

Reply:
Done.